lib/db/mysql/index.js

```diff
@@ -2,85 +2,103 @@
-      }
+  patcher.patch(function(err) {
+    if (err) {
+      logger.error('updateDbSchema', err);
+      return d.reject(err);
+    }
+    d.resolve();
+  });
 
-      logger.verbose('changeUser');
-      client.changeUser({
-        user: options.user,
-        password: options.password,
-        database: database
-      }, function(err) {
-        if (err) {
-          logger.error('changeUser', err);
-          return d.reject(err);
-        }
-        logger.verbose('creatingSchema');
+  return d.promise;
+}
+
+// Sanity-check that we're working with a compatible patch level.
+
+function checkDbPatchLevel(patcher) {
+  logger.verbose('checkDbPatchLevel', patcher.options);
+
+  var d = P.defer();
+
+  patcher.readDbPatchLevel(function(err) {
+    if (err) {
+      logger.error('checkDbPatchLevel', err);
+      return d.reject(err);
+    }
+    // We are only guaranteed to run correctly if we're at the current
+    // patch level for this version of the code (the normal state of
+    // affairs) or the one immediately above it (during a deployment).
+    if (patcher.currentPatchLevel !== patch.level) {
+      if (patcher.currentPatchLevel !== patch.level + 1) {
+        err = 'unexpected db patch level: ' + patcher.currentPatchLevel;
+        logger.error('checkDbPatchLevel', err);
+        return d.reject(new Error(err));
+      }
+    }
+    d.resolve();
+  });
 
-        client.query(SCHEMA, function(err) {
-          if (err) {
-            logger.error('creatingSchema', err);
-            return d.reject(err);
-          }
-          d.resolve();
-        });
-      });
-    });
   return d.promise;
 }
 
 MysqlStore.connect = function mysqlConnect(options) {
-  if (options.createSchema) {
-    // ugly, but you can't connect to a database before the database actually
-    // exists. So remove and restore it later.
-    var database = options.database;
-    delete options.database;
-
-    options.multipleStatements = true;
-    var schemaConn = mysql.createConnection(options);
-    options.database = database;
-
-    return createSchema(schemaConn, options).then(function() {
-      schemaConn.end();
-      delete options.multipleStatements;
-      options.database = database;
-      return new MysqlStore(options);
-    });
-  } else {
-    return P.resolve(new MysqlStore(options));
-  }
+
+  options.createDatabase = options.createSchema;
+  options.dir = path.join(__dirname, 'patches');
+  options.metaTable = 'dbMetadata';
+  options.patchKey = 'schema-patch-level';
+  options.patchLevel = patch.level;
+  options.mysql = mysql;
+  var patcher = new MysqlPatcher(options);
+
+  return P.promisify(patcher.connect, patcher)().then(function() {
+    if (options.createSchema) {
+      return updateDbSchema(patcher);
+    }
+  }).then(function() {
+    return checkDbPatchLevel(patcher);
+  }).then(function() {
+    return P.promisify(patcher.end, patcher)();
+  }).then(function() {
+    return new MysqlStore(options);
+  });
 };
```

Comment:
Not sure if you want a `.catch()` function here, just in case something goes wrong. Ah, perhaps it is checked from the calling function. Maybe not. I'm probably overlooking something.

```js
  if (config.get('db.driver') === 'mysql') {
    p = klass.connect(config.get('mysql'));
  } else {
    p = klass.connect();
  }
  return p.then(function(store) {
    logger.debug('connected', config.get('db.driver'));
    driver = store;
  }).then(loadProviders).then(function() {
    return driver;
  });
```